multibuildingdetector/evaluators/tripletevaluator.py

```diff
@@ -0,0 +1,178 @@
+        target (chainer.Link): A detection link. This link must have
+            :meth:`predict` method that takes a list of images and returns
+            following tuple :obj:`multibox_locs, multibox_triplets`.
+            :obj:`multibox_locs` is a vector containing the bbox locations
+            encoded in the default bbox space,
+            :obj:`multibox_triplets` is a vector containing the triplet loss
+            feature vectors encoded in the default bbox space.
+        label_names (iterable of strings): An iterable of names of classes.
+    """
+
+    trigger = 1, 'epoch'
+    default_name = 'validation'
+    priority = chainer.training.PRIORITY_WRITER
+
+    def __init__(
+            self, iterator, target, label_names=None,
+            save_plt=False, save_path='result'):
+        super(TripletEvaluator, self).__init__(
+            iterator, target)
+        self.label_names = label_names
+        self._save = save_plt
+        self._save_path = save_path
+
+    def evaluate(self):
+        iterator = self._iterators['main']
+        target = self._targets['main']
+
+        if hasattr(iterator, 'reset'):
+            iterator.reset()
+            it = iterator
+        else:
+            it = copy.copy(iterator)
+
+        imgs, pred_values, gt_values = apply_prediction_to_iterator(
+            target.predict, it)
+        # delete unused iterator explicitly
+        del imgs
+
+        mb_boxs, mb_confs = pred_values
+
+        _, gt_labels = gt_values
+
+        report = {}
+
+        label_groups = defaultdict(list)
+
+        for labels, confs in zip(gt_labels, mb_confs):
+            label_groups.update(SSDTripletLoss._get_label_groups(
+                self.filter_overlapping_bboxs(mb_boxs, mb_confs,
+                                              gt_labels)))
+        del label_groups[0]
+
+        self.convert_label_groups(label_groups)
+
+        if self._save:
+            self.plot_roc_curves(label_groups)
+
+        for label, feat_v in label_groups.items():
+            avg_dist = -1
+            if feat_v.shape[0] > 1:
+                distances = pdist(feat_v)
+                avg_dist = mean(distances)
+            label_name = self.label_names[label - 1]
+            report['main/avg_dist/{}'.format(label_name)] = avg_dist
+            if self._save:
+                pca = PCA(n_components=2)
+                pca_data = pca.fit_transform(feat_v)
+                plt.scatter([x[0] for x in pca_data],
+                            [x[1] for x in pca_data],
+                            label=label_name)
+        if self._save:
+            plt.legend()
+            plt.savefig(self._save_path + '/triplet_scatter.jpg')
+            plt.clf()
+        print(report)
+
+        observation = dict()
+        with reporter.report_scope(observation):
+            reporter.report(report, target)
+        return observation
+
+    def filter_overlapping_bboxs(self, mb_boxs, mb_confs, gt_labels):
```

Comment:
Maybe we can refactor the `SSDTripleloss` to make this method static callable so we don't have to add it here.